README.md
[![Build
Status](https://travis-ci.org/ices-tools-prod/msy.svg?branch=master)](https://travis-ci.org/ices-tools-prod/msy)
[![local
release](https://img.shields.io/github/release/ices-tools-prod/msy.svg?maxAge=3600)](https://github.com/ices-tools-prod/msy/tree/0.2-0)
[![CRAN
Status](http://www.r-pkg.org/badges/version/msy)](https://cran.r-project.org/package=msy)
[![CRAN
Downloads](http://cranlogs.r-pkg.org/badges/grand-total/msy)](https://cran.r-project.org/package=msy)
[![License](https://img.shields.io/badge/license-GPL%20\(%3E%3D%202\)-blue.svg)](https://www.gnu.org/licenses/gpl-3.0.en.html)

# msy

The `msy` R package is a collection of methods to estimate equilibrium
reference points for fish stocks

## Contact

You are welcome to:

  - submit suggestions and bug-reports at:
    <https://github.com/ices-tools-prod/msy/issues>
  - send a pull request on: <https://github.com/ices-tools-prod/msy>

# Preamble

This document is as much as the `msy`-package itself still in
developement.

The origin of this package is from an intial coding by John Simmonds
which was restructured by Colin Millar into an R-package with additional
development, including coding the Buckland method. Einar Hjörleifsson
compartmentalized the structure of the code as well as providing the
output of the analysis in a more structured format.

At this moment there is no person responsible for the maintenance of the
package, including the above mentioned names.

# Installation

The developmental repository for the `msy` package is located on
[github](https://github.com), more specifically on
[github.com/ices-tools-prod/msy](https://github.com/ices-tools-prod/msy)

The easiest way to install the `msy` package is to use the function
`install_github` in the `devtools` package. Since the input into the msy
is in the form of a `FLStock` object it is advised to obtain the latest
release of `FLCore` from the same site.

The [`Rtools.exe`](http://cran.at.r-project.org/bin/windows/Rtools)
software is needed for building packages under Microsoft Windows.

Run the following lines to install the latest versions of `msy` and
`FLCore`.

``` r
# install.packages("devtools")
install.packages("FLCore", repo = "http://flr-project.org/R")
library(devtools)
install_github("ices-tools-prod/msy")
```

The above is equivalent to `install.packages` and hence need only to be
performed once. However, since the `msy` package is currently under
development (including bug-fixing) one may expect more frequent code
updating in the package than what one may be familiar with for packages
on `CRAN`. Once installed the `msy` packages is simply loaded via the
familiar:

``` r
library(msy)
```

Besides functions the package comes with the following data:

  - codCS: `FLStock` object of the Celtic Sea cod
  - codEB: `FLStock` object of the Eastern Baltic cod
  - codIS: `FLStock` object of the Icelandic cod
  - codNS: `FLStock` object of the North Sea cod
  - codWB: `FLStock` object of the Western Baltic cod
  - codWS: `FLStock` object of the West of Scotland cod
  - saiFO: `FLStock` object of the Faeroe saithe
  - saiIS: `FLStock` object of the Icelandic saithe
  - solKA: `FLStock` object of the Kattegat sole

These are all stored in the icesStocks list object.

The current version of the `msy` implements two methods that go under
the working names `eqsim` and `plotMSY`. Only usage of functions for the
`eqsim` approaches are described in the following sections.

# eqsim

`eqsim` is a stochastic equilibrium software that may be used to explore
MSY reference points. Productivity parameters (i.e. year vectors for
natural mortality, weights-at-age and maturities) as well as selection
are re-sampled at random from user specified range of years from the
assessment. Fixing these parameters to an average over specified years
can also be set by the user. Recruitments are re-sampled from their
predictive distribution. Uncertainty in the stock-recruitment model is
taken into account by applying model averaging using smooth AIC weights
(Buckland et al. 1997). In addition assessment errors can be emulated by
applying a user-specified error (CV and autocorrelation) to the intended
target fishing mortality.

The current version of `eqsim` only takes `FLStock` objects as inputs.

## A quick start

In the following subsections we will simulate the north sea cod stock
into the future under some basic assumptions. For the simulations we
need to choose which years we will use to generate noise in the
quantities: weight at age, maturity at age, natural mortality at age,
and selection pattern. We also need to choose a set of Fbar values to
simulate over in order estimate F reference points.

The eqsim approach consists of three components:

1.  Estimate the stock recruitment relationship
2.  Simulate a stock to equilibrium and continue simulating for some
    years
3.  Calculate reference points from the simulated stock at equilibrium
    (last 50 years of the runs are used)

This can be done in one go with the following code:

``` r
FIT <- eqsr_fit(icesStocks$saiNS,
                nsamp = 1000,
                models = c("Ricker", "Segreg"))
SIM <- eqsim_run(FIT,
                 bio.years = c(2004, 2013),
                 sel.years = c(2004, 2013),
                 Fcv = 0.24,
                 Fphi = 0.42,
                 Blim = 106000,
                 Bpa = 200000,
                 Fscan = seq(0, 1.2, len = 40),
                 verbose = FALSE)
```

The stock recruitment function can be plotted by:

``` r
eqsr_plot(FIT,n=2e4)
```

![](README_files/figure-gfm/plot_fit-1.png)<!-- -->

Summary of the key results can be obtained
by:

``` r
SIM$Refs
```

``` 
                  F05          F10          F50    medianMSY      meanMSY
catF     3.675774e-01 3.940308e-01 4.990954e-01 3.076923e-01 3.076923e-01
lanF               NA           NA           NA 3.076923e-01 3.076923e-01
catch    1.265605e+05 1.239753e+05 9.571695e+04 1.284774e+05 1.284774e+05
landings           NA           NA           NA 1.284774e+05 1.284774e+05
catB     2.126939e+05 1.909209e+05 1.057146e+05 2.684618e+05 2.684618e+05
lanB               NA           NA           NA 2.684618e+05 2.684618e+05
             FCrash05     FCrash50
catF     4.923077e-01    0.6153846
lanF               NA           NA
catch    9.907078e+04 3703.8785837
landings           NA           NA
catB     1.113816e+05 3091.0668735
lanB               NA           NA
```

Summary plots conditioned on maximizing **catch** are obtained by:

``` r
eqsim_plot(SIM,catch=TRUE)
```

![](README_files/figure-gfm/plot_catch-1.png)<!-- -->

Summary plots of **Fmsy range** (WKMSYREF3) are obtained by:

``` r
eqsim_plot_range(SIM, type="mean")
```

![](README_files/figure-gfm/plot_msy-1.png)<!-- -->

``` r
eqsim_plot_range(SIM, type="median")
```

![](README_files/figure-gfm/plot_msy-2.png)<!-- -->

``` r
eqsim_plot_range(SIM, type="ssb")
```

![](README_files/figure-gfm/plot_msy-3.png)<!-- -->

## The recruitment model

Model fitting is done by maximum likelihood using the `nlminb` optimiser
in R. By refitting to non-parametric bootstrap resamples of the stock
and recruit pairs, samples from the approximate joint distribution of
the model parameters can be made. This is done by invoking the
`eqrs_fit` function. The function first sets up the stock and recruit
pairs based on the information in the `FLStock` object and removes any
incomplete pairs, before dispatching on the model fitting / averaging
algorithm chosen. Currently only a bootstrap based model averaging
method called smooth AIC is implemented fully. The details can be found
in `eqrs_Buckland` function. The algorithm implemented is:

1.  Take a resample with replacement from the stock and recruit pairs
2.  Fit every stock-recruit model under consideration and store the AIC
    of each
3.  Retain the parameter estimates from the best model
4.  Repeat

This process provides a robust way to average over several models, as
long as the bootstrap resampling procedure provides an adequate
approximation to the empirical distribution of the stock and recruit
pairs.

The arguments to the fitting function
    are

``` r
args(eqsr_fit)
```

    function (stk, nsamp = 5000, models = c("Ricker", "Segreg", "Bevholt"), 
        id.sr = FLCore::name(stk), remove.years = NULL) 
    NULL

Here:

  - `stk` is an `FLStock` object
  - `nsamp` is the number of simulations to run (often referred to as
    iterations)
  - `models` is the models to average over (any of the combination of
    these can be supplied, including only a single model)
  - `method` the method used (only Buckland as of now)
  - `id.sr` placeholder if one wants to name the fit
  - `remove.years` is used to remove years from the fit
  - `delta` and `nburn` are related to an MCMC based fitting procedure
    (not implemented)

The results from the fitting process are returned to the user as a list:

``` r
str(FIT, 2, give.attr=FALSE)
```

    List of 6
     $ sr.sto:'data.frame': 1000 obs. of  4 variables:
      ..$ a    : num [1:1000] 0.872 1.001 1.157 1.172 1.429 ...
      ..$ b.b  : num [1:1000] 164413 126196 106372 111533 88874 ...
      ..$ cv   : num [1:1000] 0.494 0.51 0.46 0.565 0.462 ...
      ..$ model: chr [1:1000] "Segreg" "Segreg" "Segreg" "Segreg" ...
     $ sr.det:'data.frame': 2 obs. of  6 variables:
      ..$ a    : num [1:2] 1.44 1.21
      ..$ b    : num [1:2] 3.74e-06 1.06e+05
      ..$ cv   : num [1:2] 0.521 0.492
      ..$ model: chr [1:2] "Ricker" "Segreg"
      ..$ n    : int [1:2] 142 858
      ..$ prop : num [1:2] 0.142 0.858
     $ pRec  : num [1:1000, 1:44] 93482 107228 123056 125553 127043 ...
     $ stk   :Formal class 'FLStock' [package "FLCore"] with 20 slots
     $ rby   :'data.frame': 44 obs. of  7 variables:
      ..$ year        : int [1:44] 1967 1968 1969 1970 1971 1972 1973 1974 1975 1976 ...
      ..$ rec         : num [1:44] 291836 327932 171373 152852 148741 ...
      ..$ ssb         : num [1:44] 194228 164413 263979 311949 429606 ...
      ..$ fbar        : num [1:44] 0.322 0.291 0.262 0.408 0.329 ...
      ..$ landings    : num [1:44] 113751 88326 130588 234962 265381 ...
      ..$ catch       : num [1:44] 113751 88326 130588 234962 265381 ...

NEWS
--------------------------------------------------------------------------------
msy 0.1-19 (2019-02-07)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/b11e04e61d343b6f412

--------------------------------------------------------------------------------
msy 0.1-18 (2017-02-07)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/b11e04e61d343b6f412

--------------------------------------------------------------------------------
msy 0.1-17 (2016-07-19)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/51d5e35409c606191b18ec0

--------------------------------------------------------------------------------
msy 0.1-16 (2015-03-19)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/f249f083084d16e44b9a4f1
o SR function names capitalised

--------------------------------------------------------------------------------
msy 0.1-15 (2015-03-13)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/331947083d7a613761e67

--------------------------------------------------------------------------------
msy 0.1-14 (2015-03-13)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/976d8157af52fd146

--------------------------------------------------------------------------------
msy 0.1-13 (2015-03-09)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/f4c484c23ac2d0

--------------------------------------------------------------------------------
msy 0.1-12 (2014-02-22)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/90ba8bc48f5f8b2fc4

--------------------------------------------------------------------------------
msy 0.1-11 (2013-01-24)
--------------------------------------------------------------------------------
o ...

--------------------------------------------------------------------------------
msy 0.1-10 (2014-02-02)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/ba4c2d848c1b92de2

--------------------------------------------------------------------------------
msy 0.1-9 (2014-02-02)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/75b6a6d33d83778

--------------------------------------------------------------------------------
msy 0.1-8 (2014-01-31)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/ed52ad17d63170c9

--------------------------------------------------------------------------------
msy 0.1-7 (2014-01-30)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/f33ba1379186dff69

--------------------------------------------------------------------------------
msy 0.1-6 (2014-01-30)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/a816e6acc73c4e96

--------------------------------------------------------------------------------
msy 0.1-5 (2014-01-29)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/1b6ae9b9176b2143f4

--------------------------------------------------------------------------------
msy 0.1-4 (2014-01-27)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/dba8001d634a28f


--------------------------------------------------------------------------------
msy 0.1-3 (2014-01-19)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/e4168bbd1cc1b1f


--------------------------------------------------------------------------------
msy 0.1-2 (2014-01-17)
--------------------------------------------------------------------------------
o https://github.com/ices-tools-prod/msy/commit/2b02d535941


--------------------------------------------------------------------------------
msy 0.1-1 (2013-01-24)
--------------------------------------------------------------------------------
o Initial commit


R/eqsim2.R
#' @title simulates the equilibrium results for a population
#'
#' @description XXX
#' 
#' @export
#' 
#' @author Colin Millar \email{colin.millar@@jrc.ec.europa.eu}
#' 
#' @param fit A list returned from the function fitModels
#' @param bio.years The years to sample maturity, weights and M from
#' @param bio.const A flag, if FALSE mean of the biological values from the years selected are used
#' @param sel.years The years to sample the selection patterns from
#' @param sel.const A flag, if FALSE mean of the selection patterns from the years selected are used
#' @param Fscan F values to scan over
#' @param Fcv Assessment error in the advisory year
#' @param Fphi Autocorrelation in assessment error in the advisory year
#' @param Blim This we know
#' @param Bpa This we know
#' @param recruitment.trim A numeric vector with two log-value clipping the extreme
#' recruitment values from a continuous lognormal distribution. The values must
#' be set as c("high","low").
#' @param Btrigger If other than 0 (default) the target F applied is reduced by
#' SSB/Btrigger 
#' @param Nrun The number of years to run in total (last 50 years from that will be retained)
#' @param process.error Use stochastic recruitment or mean recruitment?  (TRUE = predictive)
#' @param verbose Flag, if TRUE (default) indication of the progress of the
#' simulation is provided in the console. Useful to turn to FALSE when 
#' knitting documents.
#' @param extreme.trim Call John Simmonds :-)

eqsim_run2 <- function(fit,
                      bio.years = c(2008, 2012), # years sample weights, M and mat
                      bio.const = FALSE,
                      sel.years= c(2008, 2012), # years sample sel and discard proportion by number from
                      sel.const = FALSE,
                      Fscan = c(0.1,0.2), # F values to scan over
                      Fcv = 0,
                      Fphi = 0,
                      Blim,
                      Bpa,
                      recruitment.trim = c(3, -3),
                      Btrigger = 0,
                      n_years = 100, # number of years to run in total
                      process.error = TRUE, # use predictive recruitment or mean recruitment? (TRUE = predictive)
                      verbose = TRUE,
                      extreme.trim)
{
  
  # some minor input checking
  if (abs(Fphi) >= 1) stop("Fphi, the autocorelation parameter for log F should be between (-1, 1)")
  if ((recruitment.trim[1] + recruitment.trim[2])> 0) stop("recruitment truncation must be between a high - low range")
  
  ##############################################################################
  # 1. SETTING DIMENSIONS
  #  age
  a1 <- range(fit$stk)[1]
  a2 <- range(fit$stk)[2]
  n_ages <- dims(fit$stk)$age
  #  time
  y1 <- range(fit$stk)[5]
  y2 <- y1 + n_years
  n_years_keep <- min(n_years, 50)
  #  target
  HRATE <- Fscan
  n_target <- length(HRATE)
  #  iterations
  iter <- nrow(fit$sr.sto)
  
  ##############################################################################
  # 2. SETTING UP OBJECTS
  
  # Array of n_years and n_iter
  x <- array(0, c(n_years,iter),dimnames=list(year=1:n_years,iter=1:iter))
  SSBy <- Fy_error <- x
  
  # Array of n_ages, n_years and n_iter
  x <- array(0, c(n_ages, n_years, iter),dimnames=list(age=a1:a2,year=1:n_years,iter=1:iter))
  Nay <- Fay  <- Cay <- cWay <- lWay <- lRay <- x
  
  # TODO per note from Carmen:
  #  NOTE: If we want Fy_error to be a stationary AR(1) process, it would make
  #        more sense to initialise Fy_error as a Normal dist with zero mean and
  #        standard deviation of AR(1) marginal distribution, i.e. standard 
  #        deviation of initial Fy_error = Fcv/sqrt(1- Fphi^2), instead of just
  #        initialising Fy_error=0
  

  # Note, we could equivalent for the other ay data
  
  # initial recruitment
  R <- mean(fit$rby$rec)
  
  x <- array(0,
             c(7, n_target),
             dimnames=list(quants=c("p025","p05","p25","p50","p75","p95","p975"),target=Fscan))
  ssbs <- cats <- lans <- recs <- x
  
  x  <- array(0,
              c(n_target, n_years_keep, iter),
              dimnames=list(fmort=Fscan,year=1:n_years_keep,iter=1:iter))
  ferr <- SSB <- CATCH <- LAND <- REC <- x
  begin <- n_years - n_years_keep + 1
  
  # New from Simmonds' 29.1.2014
  #   Residuals of SR fits (1 value per SR fit and per simulation year 
  #     but the same residual value for all Fscan values):
  SR <- fit$sr.sto
  #data <- fit $ rby[,c("rec","ssb","year")]
  resids= array(rnorm(iter*(n_years+1), 0, SR$cv),c(iter, n_years+1))
  # Limit how extreme the Rec residuals can get:
  lims = t(array(SR$cv,c(iter,2))) * recruitment.trim
  for (k in 1:iter) { resids[k,resids[k,]>lims[1,k]]=lims[1,k]}
  for (k in 1:iter) { resids[k,resids[k,]<lims[2,k]]=lims[2,k]}
  # end New from Simmonds 29.1.2014
  
  
  
  
  ##############################################################################
  # A. data input
  
  # NOTE: Should make this as a function call, i.e. independent of the rest
  
  ## A1: little helper functions
  littleHelper <- function(x,i) {
    x2 <- x
    x2[i] <- NA
    x2[] <- apply(x2,1,mean,na.rm=TRUE)
    x[i] <- x2[i]
    return(x)
  }
  
  ## A2: Biological data
  b1 <- bio.years[1]
  b2 <- bio.years[2]
  stk.winbio <- window(fit$stk, start = b1, end = b2)  
  
  west <- matrix(stock.wt(stk.winbio), ncol = b2 - b1 + 1)
  i <- west == 0
  if(any(i)) west <- littleHelper(west,i)
  weca <- matrix(catch.wt(stk.winbio), ncol = b2 - b1 + 1)
  i <- weca == 0
  if(any(i)) weca <- littleHelper(weca,i)
  wela <- matrix(landings.wt(stk.winbio), ncol = b2 - b1 + 1)
  if(any(i)) wela <- littleHelper(wela,i)
  Mat <- matrix(mat(stk.winbio), ncol = b2 - b1 + 1)
  M <- matrix(m(stk.winbio), ncol = b2 - b1 + 1)
  
  # 22.2.2014 Added weight of landings per comment from Carmen
  if (bio.const==TRUE){ # take means of wts Mat and M and ratio of landings to catch
    west[] <- apply(west, 1, mean)
    weca[] <- apply(weca, 1, mean)
    wela[] <- apply(wela, 1, mean)
    Mat[] <- apply(Mat, 1, mean)
    M[] <- apply(M, 1, mean) #me
  }
  
  rand_bio <- array(sample(1:ncol(weca), n_years * iter, TRUE), c(n_years, iter))
  cWay[] <- c(weca[, c(rand_bio)])
  lWay[] <- c(wela[, c(rand_bio)])

  
  ## B2: Fisheries data
  s1 <- sel.years[1]
  s2 <- sel.years[2]
  stk.winsel <- window(fit$stk, start = s1  , end = s2)
  
  landings <- matrix(landings.n(stk.winsel), ncol = s2 - s1 + 1)
  catch <- matrix(catch.n(stk.winsel), ncol = s2 - s1 + 1)
  sel <- matrix(harvest(stk.winsel), ncol = s2 - s1 + 1)
  Fbar <- matrix(fbar(stk.winsel), ncol = s2 - s1  + 1)
  sel <- sweep(sel, 2, Fbar, "/")
  
  if (sel.const == TRUE) { # take means of selection
    sel[] <- apply(sel, 1, mean)
    landings[]  <- apply(landings, 1, mean)
    catch[]  <- apply(catch, 1, mean)
  }

  land.cat= landings / catch  # ratio of number of landings to catch
  # TODO: Check if this is sensible
  i <- is.na(land.cat)
  if(any(i)) land.cat[i] <- 1
  
  rand_sel <- array(sample(1:ncol(sel), n_years * iter, TRUE), c(n_years, iter))
  lRay[]  <- c(land.cat[, c(rand_sel)])
  
  # No variability assumed in the pFa and pMa. Here generate an age vector
  pFa <- apply(harvest.spwn(stk.winsel), 1, mean)[drop=TRUE] # vmean(harvest.spwn(stk.win))
  pMa <- apply(m.spwn(stk.winsel), 1, mean)[drop=TRUE] # mean(m.spwn(stk.win))
  
  ##############################################################################
  # B. Simulation setup
  
  if (verbose) loader(0)
  
  # Looping over each F value in Fscan. For each of the iter SR fits 
  # (replicates), do a forward simulation during n_years years
  # There are Rec residuals for each SR fit and year, which take the same
  # values for all Fscan 
  for (i in 1:n_target) {
    
    # The F value to test
    Fbar <- Fscan[i]
    
    ############################################################################
    # Population in simulation year 1:
    
    # Zpre: Z that occurs before spawning
    Zpre <- ( sel[,rand_sel[1,]]*Fbar * pFa + M[,rand_bio[1,]] * pMa)
    
    # Zpos: Z that occurs after spawning
    # Zpos not used anywhere
    # Zpos <- (Fbar * (1-pFa) * sel[,rand_sel[1,]] + M[,rand_bio[1,]] * (1-pMa))
    
    # run Z out to age 50 ...
    # TODO:
    # Comments from Carmen: Zcum is a cumulative sum, but it is done in a strange way:
    #  There is a matrix of F-at-age and a matrix of M-at-age (each has 49 ages, iter replicates)
    #  The F and M matrices are summed, giving Z-at-age (49 ages, iter replicates)
    #  But then a cumsum is taken considering the Z-at-age matrix as a vector (i.e. not column-wise) ????
    #  This is strange, by applying "cumsum" treating Z-at-age as a vector, really only the first 50 values of
    #  the resulting "Zcum" make sense (all other values seem "wrong", or at least, meaningless)
    Zcum <- c(0, cumsum(Fbar * sel[c(1:n_ages, rep(n_ages, 49 - n_ages)), rand_sel[1,]] + M[c(1:n_ages, rep(n_ages, 49 - n_ages)), rand_bio[1,]]))
    # Carmen: Following from "Zcum", only first 50 elements of N1 make sense ????
    N1 <- R * exp(- unname(Zcum))
    
    # set up age structure in first year for all simulations
    # Comments from Carmen:
    #   Nay has dimension = (no. ages, no. simulation yrs "n_years", no. SR fits "iter")
    #   With this code, we seem to be getting always the same population-at-age value for year 1
    #   instead of iter different values, as might have been intended ????
    #   (the whole problem is coming from Zcum ==> N1 ==> Nay[,1,] )
    Nay[,1,] <- c(N1[1:(n_ages-1)], sum(N1[n_ages:50]))
    
    # calculate ssb in first year using a different stock.wt and Mat selection and M for each simulation
    # Comments from Carmen:
    #   SSBy has dimension = (no. simul yrs "n_years", no. SR fits "iter")
    #   SSB in year 1:
    #   although Nay[,1,] has dim no.ages x iter, all iter values of Nay[,1,] are
    #   the same (because of Zcum issue)
    SSBy[1,] <- colSums(Mat[,rand_bio[1,]] * Nay[,1,] * west[,rand_bio[1,]] / exp(Zpre)[])
    
    # Years 2 to n_years:
    for (j in 2:n_years) {
      # get ssb from previous year
      SSBy_1 <- SSBy[j-1,]
      
      # predict recruitment using various models
      if (process.error) {
        # Changes 29.1.2014
        # new random draws each time
        # allrecs <- sapply(unique(SR $ mod), function(mod) exp(match.fun(mod) (SR, SSBy_1) + rnorm(iter, 0, SR $ cv)))
        # same random draws used for each F
        allrecs <- sapply(unique(SR$mod), function(mod) exp(match.fun(mod)(SR, SSBy_1) + resids[,j]))
        # end Changes 29.1.2014
      } else {
        allrecs <- sapply(unique(SR $ mod), function(mod) exp(match.fun(mod) (SR, SSBy_1)))
      }
      
      # Comment from Carmen:
      #  For each of the iter replicates, this selects the appropriate SR model
      #   type to use in that replicate
      #  Note that the order of SR model types that comes out in "select" is
      #   not necessarily the same order in which the SR model types were
      #   entered as inputs -- I presume the **next 2 lines** of code have
      #   been checked to avoid potential bugs due to this reordering  ???? 
      select <- cbind(seq(iter), as.numeric(factor(SR $ mod, levels = unique(SR $ mod))))
      Nay[1,j,] <- allrecs[select]
      
      # Comment from Carmen:
      #   Note: it seems that Rec is coded as occurring always at age 1
      #   (i.e. based on SSBy_1 in previous year)
      #   Some stocks have Rec at ages other than 1 (e.g. age 0) 
      #    -- is this a problem ????
      
      # apply HCR
      # (intended) Fbar to be applied in year j-1 (depends on SSBy_1 in year j-1):
      Fnext <- Fbar * pmin(1, SSBy_1/Btrigger)
      
      # apply some noise to the F
      # Notes from Carmen:
      #  Assessment and/or implementation error (modifies intended F to get
      #  realised F)
      #  Error: AR(1) process on log(F) with autocorrel = Fphi, and
      #  conditional stand deviation = Fcv
      #  Might make more sense to have the "Fy_error" matrix calculated before
      #  the Fscan loop starts so that the same errors in F are applied to
      #  all Fscan values ???? (as for Rec residuals)
      Fy_error[j,] <- Fphi * Fy_error[j-1,] + rnorm(iter, 0, Fcv)
      
      # realised Fbar in year j-1:
      Fnext <- exp(Fy_error[j,]) * Fnext
      
      # get a selection pattern for each simulation and apply this to get N
      Zpre <- rep(Fnext, each = length(pFa)) * pFa * sel[, rand_sel[j,]] + M[, rand_bio[j,]] * pMa
      
      # get Fay
      Fay[ , j-1, ] <- rep(Fnext, each = n_ages) * sel[, rand_sel[j-1,]]
      
      Nay[ -1, j, ] <- Nay[1:(n_ages-1), j-1, ] * exp(-Fay[1:(n_ages-1), j-1, ] - M[1:(n_ages-1), rand_bio[j-1,]])
      Nay[n_ages, j, ] <- Nay[n_ages, j, ] + Nay[n_ages, j-1, ] * exp(-Fay[n_ages, j-1, ] - M[n_ages, rand_bio[j-1,]])
      # calculate ssb and catch.n
      SSBy[j, ] <- apply(array(Mat[, rand_bio[j,]] * Nay[,j,] * west[, rand_bio[j,]] / exp(Zpre), c(n_ages, iter)), 2, sum)
      Cay[, j, ] <- Nay[, j-1, ] * Fay[, j-1, ] / (Fay[, j-1, ] + M[, rand_bio[j-1,]]) * (1 - exp(-Fay[, j-1, ] - M[, rand_bio[j-1,]]))
    }
    
    # convert to catch weight
    Cw <- Cay * cWay   # catch Numbers *catch wts
    land <- Cay*lRay*lWay # catch Numbers * Fraction (in number) landed and landed wts
    Lan=apply(land,2:3,sum)
    Cat <- apply(Cw, 2:3, sum)
    
    # summarise everything and spit out!
    quants <- c(0.025, 0.05, 0.25, 0.5, 0.75, 0.95, 0.975)
    ssbs[, i] <- quantile(SSBy[begin:n_years, ], quants)
    cats[, i] <- quantile(Cat[begin:n_years, ], quants)
    lans[, i] <- quantile(Lan[begin:n_years, ], quants)
    recs[, i] <- quantile(Nay[1, begin:n_years, ], quants)
    
    
    ferr[i, , ] <- Fy_error[begin:n_years, ]
    SSB[i, , ] <- SSBy[begin:n_years, ]
    CATCH[i, , ] <- Cat[begin:n_years, ]
    LAND[i, , ] <- Lan[begin:n_years, ]
    REC[i, , ] <- Nay[1, begin:n_years, ]
    
    if (verbose) loader(i/n_target)
  }
  

  
  rbp2dataframe <- function(x,variable) {
    x <- data.frame(t(x))
    x$variable <- variable
    x$Ftarget <- as.numeric(row.names(x))
    rownames(x) <- NULL
    return(x)
  }
  rbp <- rbind(rbp2dataframe(recs,"Recruitment"),
               rbp2dataframe(ssbs,"Spawning stock biomass"),
               rbp2dataframe(cats,"Catch"),
               rbp2dataframe(lans,"Landings"))
  rbp <- rbp[,c(9,8,1:7)]
  
  # STOCK REFERENCE POINTS
  
  FCrash05 <- Fscan[which.max(cats[2,]):n_target][ which(cats[2, which.max(cats[2,]):n_target] < 0.05*max(cats[2,]) )[1] ]
  FCrash50 <- Fscan[which.max(cats[4,]):n_target][ which(cats[4, which.max(cats[4,]):n_target] < 0.05*max(cats[4,]) )[1] ]
  
  
  # Einar amended 30.1.2014
  if(missing(extreme.trim)) {
    catm <- apply(CATCH, 1, mean)
    lanm <- apply(LAND, 1, mean)
  } else {
    x <- CATCH
    i <- x > quantile(x,extreme.trim[2]) |
      x < quantile(x,extreme.trim[1])
    x[i] <- NA
    catm <- apply(x, 1, mean, na.rm=TRUE)
    
    x <- LAND
    i <- x > quantile(x,extreme.trim[2]) |
      x < quantile(x,extreme.trim[1])
    x[i] <- NA
    lanm <- apply(x, 1, mean, na.rm=TRUE)
  }
  
  # end Einar amended 30.1.2014
  
  maxcatm <- which.max(catm)
  maxlanm <- which.max(lanm)
  
  # Einar added 29.1.2014
  rbp$Mean <- NA
  rbp$Mean[rbp$variable == "Catch"] <- catm
  rbp$Mean[rbp$variable == "Landings"] <- lanm
  # end Einar added 29.1.2014
  
  
  catsam <- apply(CATCH, c(1,3), mean)
  lansam <- apply(LAND, c(1,3), mean)
  maxpf <- apply(catsam, 2, which.max)
  maxpfl <- apply(lansam, 2, which.max)
  
  FmsyLan <- Fscan[maxpfl]
  msymLan <- mean(FmsyLan)
  vcumLan <- median(FmsyLan)
  fmsy.densLan <- density(FmsyLan)
  vmodeLan <- fmsy.densLan$x[which.max(fmsy.densLan$y)]
  
  FmsyCat <- Fscan[maxpf]
  msymCat <- mean(FmsyCat)
  vcumCat <- median(FmsyCat)
  fmsy.densCat <- density(FmsyCat)
  vmodeCat <- fmsy.densCat$x[which.max(fmsy.densCat$y)]
  
  pFmsyCat  <- data.frame(Ftarget=fmsy.densCat$x,
                          value=cumsum(fmsy.densCat$y * diff(fmsy.densCat$x)[1]),
                          variable="pFmsyCatch")
  pFmsyLan  <- data.frame(Ftarget=fmsy.densLan$x,
                          value=cumsum(fmsy.densLan$y * diff(fmsy.densLan$x)[1]),
                          variable="pFmsyLandings")
  pProfile <- rbind(pFmsyCat,pFmsyLan)
  
  # PA REFERENCE POINTS
  if(!missing(Blim)) {
    pBlim <- apply(SSB > Blim, 1, mean)
    
    i <- max(which(pBlim > .95))
    grad <- diff(Fscan[i + 0:1]) / diff(pBlim[i + 0:1])
    flim <- Fscan[i] + grad * (0.95 - pBlim[i]) # linear interpolation i think..
    
    i <- max(which(pBlim > .90))
    grad <- diff(Fscan[i + 0:1]) / diff(pBlim[i + 0:1])
    flim10 <- Fscan[i]+grad*(0.9-pBlim[i]) # linear interpolation i think..
    
    i <- max(which(pBlim > .50))
    grad <- diff(Fscan[i + 0:1]) / diff(pBlim[i + 0:1])
    flim50 <- Fscan[i]+grad*(0.5-pBlim[i]) # linear interpolation i think..
    
    pBlim <- data.frame(Ftarget = Fscan,value = 1-pBlim,variable="Blim")
    pProfile <- rbind(pProfile,pBlim)
  } else {
    flim <- flim10 <- flim50 <- Blim <- NA
  }
  
  if(!missing(Bpa)) {
    pBpa <- apply(SSB > Bpa, 1, mean) 
    pBpa <- data.frame(Ftarget = Fscan,value = 1-pBpa,variable="Bpa")
    pProfile <- rbind(pProfile,pBpa)
  } else {
    Bpa <- NA
  }
  
  # GENERATE REF-TABLE 
  catF <- c(flim, flim10, flim50, vcumCat, Fscan[maxcatm], FCrash05, FCrash50)
  lanF <- c(   NA,    NA,     NA, vcumLan, Fscan[maxlanm],       NA,       NA)
  catC <- approx(Fscan, cats[4,], xout = catF)$y
  lanC <- approx(Fscan, lans[4,], xout = lanF)$y
  catB <- approx(Fscan, ssbs[4,], xout = catF)$y
  lanB <- approx(Fscan, ssbs[4,], xout = lanF)$y
  
  Refs <- rbind(catF, lanF, catC, lanC, catB, lanB)
  rownames(Refs) <- c("catF","lanF","catch","landings","catB","lanB")
  colnames(Refs) <- c("Flim","Flim10","Flim50","medianMSY","meanMSY","FCrash05","FCrash50")
  
  #TODO: id.sim - user specified.
  
  return(list(ibya=list(Mat = Mat, M = M, pFa = pFa, pMa = pMa, 
                        west = west, weca = weca, sel = sel),
              rbya=list(ferr=ferr),
              rby=fit$rby, rbp=rbp, Blim=Blim, Bpa=Bpa, Refs = Refs,
              pProfile=pProfile,id.sim=fit$id.sr))
  
}